---
title: "figure2"
author: "Vanessa Dumeaux"
date: "12/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(VISION)
library(here)
library(ggplot2)
library("LaCroixColoR")
library(reshape2)
library(viridis)
library(viridisLite)
library(RColorBrewer)
library(here)
library(pheatmap)
library(dplyr)


source(here("src", "bresat.R"))

vis.pth <- here("data", "data04-VISION")
scvi.pth <- here("data", "data03-scvi")
    
ut <- readRDS(file.path(vis.pth, "ut_0-adjLogcounts_vis.rds"))

counts <- read.csv(file.path(scvi.pth, "ut_0-adjLogcounts_magic.csv"))
counts <- subset(counts, select = -c(X))
dim(counts)

ut.latent <- ut@LatentSpace
ut.metadata <- ut@metaData
sig.scores <- ut@SigScores
```

# Identify most variable VISION signatures and explore gene expression ------------

## Compute median signature score for each cluster and identify the most variable signatures across clusters

```{r}
library(dplyr)

sigtb <- data.frame(sig.scores)

df <- t(sigtb)

bs <- sig.ranksum(df, ns=1:nrow(df), full.return = TRUE)

bs.var <- apply(bs$dat, 1, var)

sel <- rownames(bs$dat)[bs.var > 0.1]
sel

# [1] "hsp2"                  "carb_reserve_metab"    "glycolysis"            "iESR_carbohydrate"    
#  [5] "heat_inducible"        "iESR_hsp"              "m_gresham"             "cc_gresham"           
#  [9] "alcohol_dehydrogenase" "rdn"                   "hyperosm_stress"       "s_histone"            
# [13] "membrane"              "rESR_RP"               "citrtic_acid_cycle"    "oxidative2"   
```

## Check consistency of expression for most variable genes in each signature across UT cells

### Plot correlation matrix

```{r}

library(MetBrewer)

breaksList = seq(-1, 1, by = 0.1)
sel.sig <- gsub("_median", "", sel)

pdf(here("exp", "exp06-figure2", "output", "siggenes_cormatrix_ut.pdf"))

plot.cc <- lapply(sel.sig, function(x){
  
  corm <- NULL
  
  magic.cellcycle <- t(counts[, colnames(counts) %in% names(ut@sigData[[x]]@sigDict), drop = FALSE])
  
  if(nrow(magic.cellcycle) > 0){
    
    colnames(magic.cellcycle) <- paste0("cell", 1:5062)
    sel.var <- which(apply(magic.cellcycle,1, var)>0.03)
    
    magic.cellcycle <- magic.cellcycle[sel.var, , drop = FALSE]
    
    if(nrow(magic.cellcycle) > 1){
      
      corm <- cor(t(magic.cellcycle))
      
      pheatmap(corm,
             color = rev(met.brewer("Cassatt1",n=21,type="continuous")), 
             cluster_rows = TRUE, cluster_cols = TRUE, main = x, breaks = breaksList,
             scale = "none")
    }
    }
})
dev.off()

```

### Check expression of the most variable genes in each signature
```{r}
cols <- as.character(lacroix_palette("PassionFruit", type = "continuous", n=3))
names(cols) <- levels(ut.metadata$leiden_scVI)

pdf(here("exp", "exp06-figure2", "output", "siggenes_heatmap_ut.pdf"))
plot.cc <- lapply(sel.sig, function(x){
      bs <- list()
      
      magic.cellcycle <- t(counts[, colnames(counts) %in% names(ut@sigData[[x]]@sigDict), drop = FALSE])
      
      if(nrow(magic.cellcycle) > 0){
        colnames(magic.cellcycle) <- paste0("cell", 1:5062)
        sel.var <- which(apply(magic.cellcycle, 1, var)>0.03)
        
        magic.cellcycle <- magic.cellcycle[sel.var, , drop = FALSE]
        rownames(ut.metadata) <- colnames(magic.cellcycle)
    
        if(nrow(magic.cellcycle) > 1){
          bs <- sig.ranksum(magic.cellcycle, ns=1:nrow(magic.cellcycle), full.return = TRUE)
          
          pheatmap(bs$dat, 
             annotation_col = ut.metadata[bs$pat.order, colnames(ut.metadata)=="leiden_scVI",  drop = FALSE], 
             annotation_colors = list(leiden_scVI=cols), 
             color = viridisLite::viridis(n=100, option="G", direction = -1), 
             cluster_rows = FALSE, cluster_cols = FALSE, main = x,
             show_colnames = FALSE, scale = "row")
          }
      }
      
      return(bs)
      })
dev.off()

names(plot.cc) <- sel.sig

names(plot.cc)[unlist(lapply(plot.cc, function(x){length(x)==0}))] # check signatures with <= 1 most variable genes
```

# Select final signatures and make figures ----------------

We selected the most variable signatures across clusters which also avoid redundancy and exhibit robustness in the directionality of expression for the most variable genes in signature.
```{r}
sel.paper <- c("oxidative2",
               "rESR_RP","s_histone",
               "m_gresham", "hsp2", "glycolysis", "rESR_RP")

sel.paper <- sel
```


## Figure 2A,B
```{r}

sigtb <- t(data.frame(sig.scores[,sel]))

bs2 <- sig.ranksum(sigtb, ns=1:nrow(sigtb), full.return = TRUE)

test <- melt(bs2$dat)

cluster.ordered <- ut.metadata[bs$pat.order, colnames(ut.metadata)=="leiden_scVI",  drop = FALSE]
cluster.ordered$Var2 <- rownames(cluster.ordered)
cluster.ordered$Var2 <- as.numeric(cluster.ordered$Var2)
test <- left_join(test, cluster.ordered)

p.rev <- ggplot(subset(test, Var1 %in% c("s_histone", "hsp2", "m_gresham", "glycolysis", "rESR_RP", "oxidative2")), aes(y=value, x=Var2, col = value))+
geom_point(size = 0.2)+
scale_color_viridis_c()+
facet_wrap(~Var1, ncol=1, scales = "free")+
theme_bw()


ggsave(here("exp", "exp06-figure2", "output", "figure2rev_signatures.pdf"), p.rev, width = 4, height = 8)

```



# Plot hsp70 ttr1 single-cell expression ------------


```{r}
hsp70 <- counts[, colnames(counts)=="HSP70"]
summary(log2(hsp70))
hist(log2(hsp70))

ttr1 <- counts[, colnames(counts)=="TTR1"]
summary(log2(ttr1))
hist(log2(ttr1))

sub.counts <- counts[colnames(counts) %in% c("TTR1", "HSP70")]

bs <- sig.ranksum(t(sub.counts), ns=1:2, full.return = TRUE)
          
pdf("output/ttr1_hsp70_heatmap_ut.pdf")
pheatmap(bs$dat, color = viridisLite::viridis(n=100, option="G", direction = -1), 
             cluster_rows = FALSE, cluster_cols = FALSE, 
             show_colnames = FALSE, scale = "row")
dev.off()
```
## Figure 2C

```{r}
hsp70high_ttr1low <- log2(hsp70) > 0.2 & log2(ttr1) <= 0.2
hsp70mid_ttr1mid <- log2(hsp70) > 0.2 & log2(ttr1) > 0.2
hsp70low_ttr1high <- log2(hsp70) <= 0.2 & log2(ttr1) > 0.2
hsp70low_ttr1low <- log2(hsp70) <= 0.2 & log2(ttr1) <= 0.2

color_point <- c("#4daf4a", #green
                 "#e41a1c", #red
                 "grey",
                 "black"#blue
                )
group <- rep("hsp70low_ttr1low", length(hsp70))
group <- ifelse(hsp70high_ttr1low==TRUE, "hsp70high_ttr1low", group)
group <- ifelse(hsp70mid_ttr1mid==TRUE, "hsp70mid_ttr1mid", group)
group <- ifelse(hsp70low_ttr1high==TRUE, "hsp70low_ttr1high", group)



library(ggplot2)
library(ggExtra)
df <- data.frame(hsp70 = hsp70,
                 ttr1 = ttr1,
                  group = group)

p6 <- ggplot(df, aes(y=ttr1, x=hsp70, color = group))+
  geom_point(size = 0.5) +
  scale_color_manual(values = color_point, guide=FALSE) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), axis.title=element_blank(),legend.position="none")
p6 <- ggMarginal(p6, type = "histogram")

ggsave(here("exp", "exp06-figure2", "output", "figure2c-ttr1_hsp70_scatterplot.pdf"), p6, width = 4, height = 3)

```

```{r}
table(group)

high.hsp70 <- log2(hsp70) > 0.2
high.ttr1 <- log2(ttr1) > 0.2

table(high.hsp70, high.ttr1)

mcnemar.test(table(high.hsp70, high.ttr1))
```


# Plot hsp70 ttr1 microscopy ------------

## Import data

```{r}
datadir <- here("data", "microscopy", "csvs")

files <- list.files(datadir, full.names = TRUE, recursive = TRUE)
files <- files[grep("day0", files)]


all.df.list <- lapply(files, function(x){
  df <- read.csv(x)
  df <- df[,-1]
  df[["channel"]] <- c("R", "G", "B")
  
  tb <- melt(df)

  filename <- unlist(lapply(strsplit(x, "/"), "[", 10))
  foldername <- unlist(lapply(strsplit(x, "/"), "[", 9))
  tb$drug.series <- unlist(lapply(strsplit(foldername, "-"), "[", 2))
  image <- unlist(lapply(strsplit(filename, "-"), "[", 3))
  cell <- unlist(lapply(strsplit(as.character(tb$variable), "Area|Min|Max|Mean"), "[", 2))
  tb$image <- paste0(tb$drug.series, image)
  tb$cell <- paste0(tb$image, cell, sep="_")
  tb$type <- "Area"
  tb$type[grep("Mean", tb$variable)] <- "Mean"
  tb$type[grep("Min", tb$variable)] <- "Min"
  tb$type[grep("Max", tb$variable)] <- "Max"
  
  tb$measure <- paste0(tb$type, tb$channel)
  return(tb)
})

all.df <- plyr::ldply(all.df.list)
```

```{r}
wide <- all.df %>%
  select(cell, drug.series, image, value, measure) %>%
  tidyr::pivot_wider(names_from = measure, values_from = value)
```

## Explore the two experimental batches
```{r}
ggplot(wide, aes(x=1:nrow(wide),y=MeanG, color = image)) +
  geom_point(size = 0.5) + 
  facet_wrap(~drug.series, scales = "free") +
  theme(legend.position = "none")

ggplot(wide, aes(x=1:nrow(wide),y=MeanR, color = image)) +
  geom_point() + 
  facet_wrap(~drug.series, scales = "free") +
  theme(legend.position = "none")
```
## Exlcude outlier measurements and scale within each experimental batch
```{r}

wide.noOutlier <- wide[wide$MeanG < 2000 & wide$MeanR<300, ]

wide.noOutlier <- wide.noOutlier %>%
  group_by(drug.series) %>%
  mutate(scale_meang = scale(MeanG),
         scale_meanr = scale(MeanR))


ggplot(wide.noOutlier, aes(x=1:nrow(wide.noOutlier),y=scale_meang, color = image)) +
  geom_point() + 
  facet_grid(~drug.series, scales = "free") +
  theme(legend.position = "none")

ggplot(wide.noOutlier, aes(x=1:nrow(wide.noOutlier),y=scale_meanr, color = image)) +
  geom_point() + 
  facet_grid(~drug.series, scales = "free") +
  theme(legend.position = "none")
```
## Figure 2E

```{r}
summary(wide.noOutlier$scale_meanr)
summary(wide.noOutlier$scale_meang)


gth <- 0.3
rth <- 0.45


wide.noOutlier$group <- "hsp70low_ttr1low"
wide.noOutlier$group <- ifelse(wide.noOutlier$scale_meanr<=rth & wide.noOutlier$scale_meang>gth, "hsp70high_ttr1low", wide.noOutlier$group)
wide.noOutlier$group <- ifelse(wide.noOutlier$scale_meanr>rth & wide.noOutlier$scale_meang<=gth, "hsp70low_ttr1high", wide.noOutlier$group)
wide.noOutlier$group <- ifelse(wide.noOutlier$scale_meanr>rth & wide.noOutlier$scale_meang>gth, "hsp70high_ttr1high", wide.noOutlier$group)



color_point <- c("black", #blue
                 "#4daf4a", #green
                 "#e41a1c", #red
                 "grey"
                )

p7 <- ggplot(wide.noOutlier, aes(x=scale_meang, y=scale_meanr, color=group))+
    geom_point(size = 0.5) +
  scale_color_manual(values = color_point, guide=FALSE) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"), axis.title=element_blank(),legend.position="none")

p7 <- ggMarginal(p7, type = "histogram")
p7

ggsave(here("exp", "exp06-figure2", "output", "figure2e-ttr1_hsp70_microscopyscatterplot.pdf"), p7, width = 4, height = 3)

high.hsp70 <- wide.noOutlier$scale_meang > gth
high.ttr1 <- wide.noOutlier$scale_meanr > rth

table(high.hsp70, high.ttr1)

mcnemar.test(table(high.hsp70, high.ttr1))

```



