---
title: "figure1"
author: "Vanessa Dumeaux"
date: "4/21/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(ggplot2)
library(dplyr)
# devtools::install_github("johannesbjork/LaCroixColoR")
library("LaCroixColoR")
library(here)
library(ggplot2)
library(edgeR)

scvi.pth <- here("data", "data03-scvi")
```


```{r}
clusters <- read.csv(file.path(scvi.pth, "finalall-adjLogcounts_labels.csv"))
clusters <- subset(clusters,select = -c(X))
clusters$day <- as.factor(clusters$day)
clusters$batch <- as.factor(clusters$batch)
clusters$nruns <- as.factor(clusters$nruns)

table(clusters$orig.ident)
```
# Figure 1B
```{r}

clusters$orig.ident <- factor(clusters$orig.ident, 
                              levels = c("FCZ_6_3_FALSE",
                                         "FCZ_3_3_FALSE", "FCZ_3_2_FALSE", "FCZ_3_1_FALSE",
                                         "FCZ_2_2_FALSE", "FCZ_2_1_FALSE",
                                         "CSP_2_1_FALSE",  "RAPA_2_2_FALSE",
                                         "UT_0_3_FALSE", "UT_0_2_FALSE", "UT_0_1_FALSE"), ordered = TRUE)

p1 <- ggplot(clusters, aes(y=orig.ident, x=n_counts, col = orig.ident)) +
  geom_violin() + 
  geom_point() +
  scale_y_discrete(limits = levels(clusters$orig.ident))+
  scale_x_continuous(name="read counts (n)")+
  scale_color_manual(values=c(lacroix_palette("Coconut", type = "continuous", n=21)[3:8], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[13:15])) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
        axis.title.y=element_blank(), legend.position="none")

ggsave("output/figure1b-ncounts_samples.pdf", p1, width = 4,  height = 4)

pbar<- clusters %>%
  group_by(orig.ident) %>%
  summarise(n= n())

p2 <- ggplot(pbar, aes(x=orig.ident, y=n, fill = orig.ident)) +
  geom_bar(stat ="identity") +
  scale_y_continuous(name="cells (n)")+
  scale_fill_manual(values=c(lacroix_palette("Coconut", type = "continuous", n=21)[3:8], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[13:15])) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
        axis.title=element_blank(),legend.position="none") +
  coord_flip()


ggsave("output/figure1b-ncells_samples.pdf", p2, width = 3, height = 4)
```

# Figure 1 - figure supplement 2B
```{r}

p3 <- ggplot(clusters, aes(y=orig.ident, x=n_genes, col = orig.ident)) +
  geom_violin() + 
  geom_point() +
  scale_x_continuous(name="n genes")+
  scale_y_discrete(limits = levels(clusters$orig.ident))+
  scale_color_manual(values=c(lacroix_palette("Coconut", type = "continuous", n=21)[3:8], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[13:15])) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
        axis.title.y=element_blank(),legend.position="none")

ggsave("output/fig1supp2b-ngenes_samples.pdf", p3, width = 4,  height = 4)

```

# Figure 1 - figure supplement 2C

Here we need to load the data before gene filtering

```{r}
sce <-  readRDS(here("data", "data02-emptyDrops_sceQC", "filtered-sce.rds"))

genesincells <- apply(assay(sce)>0, 1, sum)

df.plot <- data.frame(value = genesincells)

threshold <- 20 # show threshold we use to exclude genes expressed in less than 20 cells

library(ggplot2)
p4 <- ggplot(df.plot, aes(log2(value))) +
  geom_histogram() +
  theme_bw() +
  geom_vline(xintercept = log2(20), color = "red", linetype = "dotted")


ggsave("output/fig1supp2c-genesincells.pdf", p4, width = 3.5, height = 3)
```

# Figure 1 - figure supplement 3A

## Build filtered pseudobulk 

```{r}
counts <- read.csv(file.path(scvi.pth, "finalall_raw.csv"), row.names = 1)
counts <- t(counts)


clusters <- read.csv(file.path(scvi.pth, "finalall-adjLogcounts_labels.csv"))



pseudo.cpm <- lapply(unique(clusters$drug_day), function(x){
    sub.sce.counts <- counts[, clusters$drug_day==x]
    df <- data.frame(rowSums(sub.sce.counts))
    df <- edgeR::cpm(df)
    df <- log2(df+1)
    colnames(df) <- x
    return(df)
  })


pseudo.cpm <- do.call(cbind,pseudo.cpm)

```

```{r}

pseudo.cpm1 <- pseudo.cpm %>% as.data.frame() %>%
  tibble::rownames_to_column()
pseudo.fcz2 <- pseudo.cpm1[, grep("FCZ_2|FCZ_3|rowname", colnames(pseudo.cpm1))]

rownames(pseudo.fcz2) <- pseudo.fcz2$rowname

p.fcz23 <- GGally::ggpairs(pseudo.fcz2[,-1], xlab = "log2(cpm+1)", ylab = "log2(cpm+1)") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), strip.background = element_blank(),
        panel.background = element_blank(), axis.line = element_line(colour = "black"),
        legend.position="none")

ggsave("output/figure1supp3a-pseudo_corrplots_fcz23.pdf", p.fcz23, width = 4, height = 4)
```

# Compare bulk and unfiltered pseudo-bulk ----

## Load bulk profile and normalize counts

```{r}
bulk.counts <- readRDS(here("data", "data01-preprocessing", "pseudobulk-star", "bulk_counts_geneid.rds"))
gtf <- as.data.frame(rtracklayer::import(here("data", "data01-preprocessing", "genome", "ca_original_genomic.gtf")))

bulk.counts <- round(bulk.counts$counts)

bulk.cpm <- edgeR::cpm(bulk.counts)
colnames(bulk.cpm) <- gsub("Aligned.sortedByCoord.out.bam", "", colnames(bulk.cpm))

gtf.sub <- unique(gtf[gtf$gene_id %in% rownames(bulk.cpm),c("gene_id", "gene_name")])

idx <- match(gtf.sub$gene_id, rownames(bulk.cpm))
gtf.sub <- gtf.sub[idx,]


rownames(bulk.cpm) <- ifelse(!is.na(gtf.sub$gene_name), gtf.sub$gene_name, rownames(bulk.cpm))
rownames(bulk.cpm) <- gsub("gene-CAALFM_", "", rownames(bulk.cpm))

# Exclude genes with 0 counts
bulk.cpm <- bulk.cpm[rowSums(bulk.cpm)!=0,] 
```

## Load unfiltered pseudo-bulk and normalize counts
```{r}
f1 <- readRDS(here("data", "data01-preprocessing", "pseudobulk-star", "pseudobulk_counts_geneid.rds"))

pseudo.counts <- round(f1$counts)

# Select FCZ samples (day2-3)
incl <- grep("FCZ_2|FCZ_3", colnames(pseudo.counts))
pseudo.cpm <- edgeR::cpm(pseudo.counts[, incl])
colnames(pseudo.cpm) <- gsub(".outAligned.sortedByCoord.out.bam", "", colnames(pseudo.cpm))
colnames(pseudo.cpm) <- paste(colnames(pseudo.cpm), "upseudo", sep = "_")

gtf.sub <- unique(gtf[gtf$gene_id %in% rownames(pseudo.cpm),c("gene_id", "gene_name")])

idx <- match(gtf.sub$gene_id, rownames(pseudo.cpm))
gtf.sub <- gtf.sub[idx,]


rownames(pseudo.cpm) <- ifelse(!is.na(gtf.sub$gene_name), gtf.sub$gene_name, rownames(pseudo.cpm))
rownames(pseudo.cpm) <- gsub("gene-CAALFM_", "", rownames(pseudo.cpm))

pseudowith0 <- rownames(pseudo.cpm)[apply(pseudo.cpm==0, 1, any)]
```


## Figure 1 - figure supplement 3B
```{r}
bulk.long <- reshape2::melt(bulk.cpm)

bulk.long$pseudowith0 <- rownames(bulk.long) %in% pseudowith0

table(bulk.long$pseudowith0)

colnames(bulk.long)[1] <- "cpm"

p <- ggplot(bulk.long, aes(x = log2(cpm+1), color = pseudowith0)) +
  geom_density() +
  theme_bw()
ggsave("output/figure1supp3b-peusobulkwith0_bulkexpression.pdf", p, height = 4, width = 5)
```

## Figure1 - figure supplement 3C

```{r}

bulk.cpm1 <- log2(bulk.cpm+1)
bulk.cpm1 <- bulk.cpm1 %>% as.data.frame() %>%
  tibble::rownames_to_column()


pseudo.cpm1 <- log2(pseudo.cpm+1)
pseudo.cpm1 <- pseudo.cpm1 %>% as.data.frame() %>%
  tibble::rownames_to_column()

fcz.cpm <- dplyr::inner_join(bulk.cpm1,pseudo.cpm1) 
rownames(fcz.cpm) <- fcz.cpm$rowname


library(psych)
pdf("output/figure1supp3c-fcz2-3_corr_upseudo-bulk.pdf")
pairs.panels(fcz.cpm[,-1], 
             method = "spearman", # correlation method
             hist.col = "#00AFBB",
             density = TRUE,  # show density plots
             ellipses = TRUE # show correlation ellipses
             )
dev.off()
```



