---
title: "figure3"
author: "Vanessa Dumeaux"
date: "12/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(VISION)
library(here)
library(ggplot2)
library("LaCroixColoR")
library(reshape2)
library(viridis)
library(viridisLite)
library(RColorBrewer)
library(here)
library(pheatmap)
library(dplyr)
library(readxl)
library(reshape2)


source(here("src", "bresat.R"))
```



# Single-cell data analyses ------------

## Import scvi, magic counts and VISION results

```{r}
vis.pth <- here("data", "data04-VISION")
scvi.pth <- here("data", "data03-scvi")
    
finalall <- readRDS(file.path(vis.pth, "utfczcsprapa-adjLogcounts_vis.rds"))

counts <- read.csv(file.path(scvi.pth, "utfczcsprapa-adjLogcounts_magic.csv"))
counts <- subset(counts, select = -c(X))
dim(counts)

all.latent <- finalall@LatentSpace
all.metadata <- finalall@metaData
sig.scores <- finalall@SigScores
```

## Figure 3A - UMAP and leiden clusters
```{r}

cl.tb <- table(all.metadata$leiden_scVI)
cl.tb

cl.sel <- names(cl.tb)[cl.tb < 200]

all.metadata$leiden_scVI_2 <- as.character(all.metadata$leiden_scVI)
all.metadata$leiden_scVI_2[all.metadata$leiden_scVI %in% cl.sel] <- "others"
all.metadata$leiden_scVI_2 <- recode_factor(all.metadata$leiden_scVI_2, "0" = "cluster1", "1" = "cluster2", "2" = "cluster3",
                                         "3" = "cluster4", "4" = "cluster5", "others" = "others")


table(all.metadata$leiden_scVI_2, all.metadata$drug_day)
table(all.metadata$leiden_scVI_2)
table(all.metadata$drug_day)

colnames(all.latent) <- c("UMAP1", "UMAP2")

full.df <- cbind(all.latent, all.metadata, sig.scores)

```

we relabel the clusters in the paper so it starts with cluster 1
leiden scvi cluster 0 = cluster 1 
leiden scvi cluster 1 = cluster 2 
leiden scvi cluster 2 = cluster 3 
leiden scvi cluster 3 = cluster 4
leiden scvi others = others



```{r}


p <- ggplot(full.df, aes(x=UMAP1, y=UMAP2, col = leiden_scVI_2)) +
  geom_point(size = 0.4) + 
  scale_color_manual(values=lacroix_palette("PassionFruit", type = "continuous", n=6)) +
  theme_void()

ggsave("output/figure3a-utfczcsprapa_umap_leiden.pdf", p, width = 4, height = 4)

```

## Figure 3B
```{r}

df.dotplot <- full.df[, c("drug_day", "leiden_scVI_2")]

df.dotplot <- df.dotplot %>%
  group_by(drug_day, leiden_scVI_2) %>%
  summarise (ct = n()) %>%
  group_by(drug_day) %>%
  mutate(freq = ct / sum(ct)) %>%
  group_by(leiden_scVI_2) %>%
  mutate(freqcluster = ct / sum(ct))



df.dotplot$drug_day <- factor(df.dotplot$drug_day, 
                             levels = rev(c("UT_0", "RAPA_2", "CSP_2", "FCZ_2", "FCZ_3")), ordered = TRUE)
df.dotplot$leiden_scVI_2 <- factor(df.dotplot$leiden_scVI_2, levels = rev(levels(df.dotplot$leiden_scVI_2)), ordered = TRUE)


p1 <- ggplot(df.dotplot, aes(y=drug_day, x = leiden_scVI_2,
                         size=freqcluster, colour = freq)) +
  geom_point() +
  coord_flip() +
  theme_minimal() +
  scale_color_viridis_c(option = "inferno") +
  theme(legend.position="top", legend.box = "vertical", legend.justification='left',
        legend.key.height= unit(0.2, 'cm'),
        legend.key.width= unit(0.2, 'cm'),
        panel.grid.major = element_line(size=0.5),
        panel.grid.minor = element_blank(),
        axis.text = element_text(size=6),
        text = element_text(size=3),
        axis.title = element_blank(),
        title =element_text(size=4))
ggsave("output/figure3b-dotplot_clusters.pdf", p1, width = 3, height = 3)


```


## Identify most variable VISION signatures and explore gene expression

### Compute median signature score for each cluster and identify the most variable signatures across clusters

```{r}
library(dplyr)

clusters <- all.metadata

sigtb <- data.frame(sig.scores)
sigtb$leiden_scvi <- clusters$leiden_scVI_2

sig.group <- sigtb %>%
  group_by(leiden_scvi) %>%
  summarise_at(.vars = names(.)[1:43],.funs = c(median="median"))

df.median.sig <- t(sig.group[, -1])
colnames(df.median.sig) <- sig.group$leiden_scvi


bs <- sig.ranksum(df.median.sig, ns=1:nrow(df.median.sig), full.return = TRUE)

bs.var <- apply(bs$dat, 1, var)

sel <- rownames(bs$dat)[bs.var > 0.05]
sel

#  [1] "iESR_carbohydrate_median"     "heat_inducible_median"        "hsp2_median"                 
#  [4] "iESR_hsp_median"              "carb_reserve_metab_median"    "alcohol_dehydrogenase_median"
#  [7] "rdn_median"                   "hyperosm_stress_median"       "hog1_UP_median"              
# [10] "glycolysis_median"            "S_G2_median"                  "S_G2_2_median"               
# [13] "oxidative2_median"            "s_histone_median"             "m_gresham_median"            
# [16] "efflux_pumps_median"          "rESR_RP_median"               "AA_biosynt_GCN4_median"
```

### Check consistency of expression for most variable genes in each signature across UT cells

#### Plot correlation matrix

```{r}

library(MetBrewer)

breaksList = seq(-1, 1, by = 0.1)
sel.sig <- gsub("_median", "", sel)

pdf("output/siggenes_cormatrix_ut.pdf")

plot.cc <- lapply(sel.sig, function(x){
  
  corm <- NULL
  
  magic.cellcycle <- t(counts[, colnames(counts) %in% names(finalall@sigData[[x]]@sigDict), drop = FALSE])
  
  if(nrow(magic.cellcycle) > 0){
    
    colnames(magic.cellcycle) <- paste0("cell", 1:16371)
    sel.var <- which(apply(magic.cellcycle,1, var)>0.03)
    
    magic.cellcycle <- magic.cellcycle[sel.var, , drop = FALSE]
    
    if(nrow(magic.cellcycle) > 1){
      
      corm <- cor(t(magic.cellcycle))
      
      pheatmap(corm,
             color = rev(met.brewer("Cassatt1",n=21,type="continuous")), 
             cluster_rows = TRUE, cluster_cols = TRUE, main = x, breaks = breaksList,
             scale = "none")
    }
    }
})
dev.off()

```

### Check expression of the most variable genes in each signature
```{r}
cols <- as.character(lacroix_palette("PassionFruit", type = "continuous", n=6))
names(cols) <- levels(all.metadata$leiden_scVI_2)

pdf("output/siggenes_heatmap_ut.pdf")
plot.cc <- lapply(sel.sig, function(x){
      bs <- list()
      
      magic.cellcycle <- t(counts[, colnames(counts) %in% names(finalall@sigData[[x]]@sigDict), drop = FALSE])
      
      if(nrow(magic.cellcycle) > 0){
        colnames(magic.cellcycle) <- paste0("cell", 1:16371)
        sel.var <- which(apply(magic.cellcycle, 1, var)>0.03)
        
        magic.cellcycle <- magic.cellcycle[sel.var, , drop = FALSE]
        rownames(all.metadata) <- colnames(magic.cellcycle)
    
        if(nrow(magic.cellcycle) > 1){
          bs <- sig.ranksum(magic.cellcycle, ns=1:nrow(magic.cellcycle), full.return = TRUE)
          
          pheatmap(bs$dat, 
             annotation_col = all.metadata[bs$pat.order, colnames(all.metadata)=="leiden_scVI_2",  drop = FALSE], 
             annotation_colors = list(leiden_scVI_2=cols), 
             color = viridisLite::viridis(n=100, option="G", direction = -1), 
             cluster_rows = FALSE, cluster_cols = FALSE, main = x,
             show_colnames = FALSE, scale = "row")
          }
      }
      
      return(bs)
      })
dev.off()

names(plot.cc) <- sel.sig

names(plot.cc)[unlist(lapply(plot.cc, function(x){length(x)==0}))] # check signatures with <= 1 most variable genes
```

# Select final signatures and make figures ----------------

We selected the most variable signatures across clusters which also avoid redundancy and exhibit robustness in the the directionality of expression for the most variable genes in signature.
```{r}
sel.paper.median <- c("hsp2_median", "rdn_median", "hyperosm_stress_median",
               "carb_reserve_metab_median", "glycolysis_median", "S_G2_median", "oxidative2_median", 
                "rESR_RP_median", "m_gresham_median",
               "efflux_pumps_median","AA_biosynt_GCN4_median")


```


## Figure 3C

```{r}

order_clusters <- c("others", "cluster1", "cluster5", "cluster3", "cluster2","cluster4")

df.plot.sig.heatmap <- df.median.sig[match(sel.paper.median, rownames(df.median.sig)), match(order_clusters, colnames(df.median.sig))]

#sel.paper <- recode_factor(sel.paper, )

pheatmap::pheatmap(df.plot.sig.heatmap,
             color = viridisLite::viridis(n=100, option="G", direction = -1), 
             cluster_rows = FALSE, cluster_cols = FALSE, 
             labels_row = c("heat-shock stress", 
                           "rRNA", "hyperosmotic stress", 
                           "carb. reserve metab.",
                           "glycolysis", 
                           "S phase", 
                           "oxidative stress",
                           "RP",
                           "M phase",
                           "efflux pumps",
                           "AA biosynth. GCN4"),
             scale = "row", filename = "output/figure3c-heatmap_visionSig_sel_bresat.pdf", height = 3, width = 4)
       

```

## Appendix2 - figure supplement 1B
```{r}


p.glyc <- ggplot(subset(full.df, !leiden_scVI_2=="others"), aes(x=UMAP1, y=UMAP2, col = glycolysis)) +
  geom_point(size = 0.4) + 
  scale_color_viridis_c()+
  theme_void()

ggsave("output/appendix2supp1b-umap_glycolysis.pdf", p.glyc, width = 4.2, height = 4)

```


## Appendix2 - figure supplement 1C

```{r}

sel.paper <- gsub("_median", "", sel.paper.median)
sel.scores <- full.df[, colnames(full.df) %in% c(sel.paper, "leiden_scVI_2", "drug_day")]

df.ecdf <- sel.scores %>% 
  reshape2::melt()

df.ecdf$variable <- factor(df.ecdf$variable, 
                           levels = c("hsp2", "rdn", "hyperosm_stress",
                                      "carb_reserve_metab", "glycolysis", "S_G2", "oxidative2",
                                      "rESR_RP", "m_gresham",
                                      "efflux_pumps","AA_biosynt_GCN4"),
                           labels = c("heat-shock stress",
                                      "rRNA", "hyperosmotic stress",
                                      "carb. reserve metab.", "glycolysis", "S phase", "oxidative stress",
                                      "RP","M phase",
                                      "efflux pumps", "AA biosynth. GCN4"),
                           ordered = TRUE)

df.ecdf$drug_day <- factor(df.ecdf$drug_day, 
                              levels = c("FCZ_3","FCZ_2", "CSP_2",  "RAPA_2","UT_0"), ordered = TRUE)


p11 <- ggplot(subset(df.ecdf, !leiden_scVI_2=="others" & variable == "glycolysis"), 
       aes(value, color = drug_day)) +
  stat_ecdf() +
  scale_color_manual(values=c(lacroix_palette("Coconut", type = "continuous", n=21)[c(1,4)], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[15])) +
  xlab("signature score") +
  ylab("ECDF") +
  facet_grid(~variable, scales = "free") +
  theme(strip.background = element_blank(),
        panel.background = element_rect(fill = "white", colour = NA), 
        panel.border = element_rect(fill = NA, colour = "grey20"), 
        panel.grid = element_line(colour = "grey92"), 
        panel.grid.minor = element_line(size = rel(0.5)), 
        legend.key = element_rect(fill = "white", colour = NA), complete = TRUE)

p11
ggsave("output/appendix2supp1c-ecdf_glycolysis_drug_day.pdf", p11, height = 4, width = 5.5)

ks.test(full.df[full.df$drug_day=="UT_0", "glycolysis"],
        full.df[full.df$drug_day=="RAPA_2", "glycolysis"])

```
## Figure 3F

```{r}
p.rp <- ggplot(subset(full.df, !leiden_scVI_2=="others"), aes(x=UMAP1, y=UMAP2, col = rESR_RP)) +
  geom_point(size = 0.4) + 
  scale_color_viridis_c()+
  theme_void()

ggsave("output/figure3f-umap_rp.pdf", p.rp, width = 4, height = 3.5)

```

## Figure 3G

```{r}
p.hsp <- ggplot(subset(full.df, !leiden_scVI_2=="others"), aes(x=UMAP1, y=UMAP2, col = hsp2)) +
  geom_point(size = 0.4) + 
  scale_color_viridis_c()+
  theme_void()

ggsave("output/figure3g-umap_hsp.pdf", p.hsp, width = 4, height = 3.5)

```

## Appendix2 - figure1G

```{r}
p.hyperosm <- ggplot(subset(full.df, !leiden_scVI_2=="others"), aes(x=UMAP1, y=UMAP2, col = hyperosm_stress)) +
  geom_point(size = 0.4) + 
  scale_color_viridis_c()+
  theme_void()

ggsave("output/appendix2figure1g-umap_hyperosm.pdf", p.hyperosm, width = 4.5, height = 3.5)

```

## Figure 3 - figure supplement 2A

```{r}

p2 <- ggplot(df.ecdf,
       aes(value, color = leiden_scVI_2)) +
  stat_ecdf() +
  scale_color_manual(values=lacroix_palette("PassionFruit", type = "continuous", n=6)) +
  facet_wrap(~variable, scales = "free_x", nrow = 2) +
  scale_x_continuous(name = "signature score") +
  scale_y_continuous(name = "ECDF") +
  theme(panel.background = element_rect(fill = "white", colour = NA), 
        panel.border = element_rect(fill = NA, colour = "grey20"), 
        panel.grid = element_line(colour = "grey92"),
        panel.grid.minor = element_line(size = rel(0.5)), 
        strip.background = element_blank(),
        legend.key = element_rect(fill = "white", colour = NA), 
        legend.title = element_blank(), complete = TRUE)

ggsave("output/figure3supp2a-ecdf_sigscores_clusters.pdf", p2, height = 4, width = 9)

```

## Appendix 2 - figure supplement 1H

```{r}
df.ecdf <- sel.scores %>% 
  reshape2::melt()

df.ecdf$variable <- factor(df.ecdf$variable, 
                           levels = c("hyperosm_stress"),
                           labels = c("hyperosmoticS"),
                           ordered = TRUE)


df.ecdf$drug_day <- factor(df.ecdf$drug_day, 
                             levels = rev(c("UT_0", "RAPA_2", "CSP_2", "FCZ_2", "FCZ_3")), ordered = TRUE)

p3 <- ggplot(subset(df.ecdf, !leiden_scVI_2=="others"),
       aes(value, color = drug_day)) +
  stat_ecdf() +
  scale_color_manual(values=c(lacroix_palette("Coconut", type = "continuous", n=21)[c(1,4)], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[15])) +
  labs (x = "signature score", y = "ECDF") +
  facet_grid(leiden_scVI_2~variable, scales = "free_x") +
  theme(strip.background = element_blank(),
        strip.text = element_text(size = 7.5), 
        axis.text = element_text(size = 7), 
        # axis.title = element_text(size = 7),
        panel.background = element_rect(fill = "white", colour = NA), 
        panel.border = element_rect(fill = NA, colour = "grey20"), 
        panel.grid = element_line(colour = "grey92"), 
        panel.grid.minor = element_line(size = rel(0.5)), 
        legend.key = element_rect(fill = "white", colour = NA), complete = TRUE)

ggsave("output/appendix2fig1h-ecdf_sigscores_drug-day.pdf", p3, height = 4)

```



## Appendix 2 - figure 1I

```{r}
magic.hyperosm <- t(counts[, colnames(counts) %in% names(finalall@sigData$hyperosm_stress@sigDict), drop = FALSE])
      
colnames(magic.hyperosm) <- paste0("cell", 1:16371)
sel.var <- which(apply(magic.hyperosm, 1, var)>0.03)

magic.hyperosm <- magic.hyperosm[sel.var, , drop = FALSE]
rownames(full.df) <- colnames(magic.hyperosm)

bs <- sig.ranksum(magic.hyperosm, ns=1:nrow(magic.hyperosm), full.return = TRUE)

col_clusters <- as.character(lacroix_palette("PassionFruit", type = "continuous", n=6))
names(col_clusters) <- levels(full.df$leiden_scVI_2)


full.df$drug_day <- factor(full.df$drug_day, 
                             levels = rev(c("UT_0", "RAPA_2", "CSP_2", "FCZ_2", "FCZ_3")), ordered = TRUE)

col_drug_day <- c(lacroix_palette("Coconut", type = "continuous", n=21)[c(1,4)], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[15])
names(col_drug_day) <- levels(full.df$drug_day)


pheatmap(bs$dat,
         annotation_col = full.df[bs$pat.order, colnames(full.df) %in% c("leiden_scVI_2", "drug_day"),  drop = FALSE],
         annotation_colors = list(leiden_scVI_2=col_clusters, drug_day = col_drug_day), 
             color = viridisLite::viridis(n=100, option="G", direction = -1), 
             cluster_rows = FALSE, cluster_cols = FALSE, main = "Hyperosmotic stress",
             show_colnames = FALSE, scale = "row", filename = "output/appendix2fig1i-hyperosm_gene_heatmap_drug_day.pdf", height = 3)
```


## Appendix2 - figure 1J
```{r}
p22 <- ggplot(subset(df.ecdf, !leiden_scVI_2=="others" & variable == "RP"), 
       aes(value, color = drug_day)) +
  stat_ecdf() +
  scale_color_manual(values=c(lacroix_palette("Coconut", type = "continuous", n=21)[c(1,4)], 
                              lacroix_palette("Coconut", type = "continuous", n=21)[10:11],
                              lacroix_palette("Coconut", type = "continuous", n=15)[15])) +
  xlab("signature score") +
  ylab("ECDF") +
  theme(strip.background = element_blank(),
        panel.background = element_rect(fill = "white", colour = NA), 
        panel.border = element_rect(fill = NA, colour = "grey20"), 
        panel.grid = element_line(colour = "grey92"), 
        panel.grid.minor = element_line(size = rel(0.5)), 
        legend.key = element_rect(fill = "white", colour = NA), complete = TRUE)

p22
ggsave("output/appendix2fig1j-ecdf_rp_drug_day.pdf", p22, height = 3.5, width = 8)

ks.test(full.df[full.df$drug_day=="UT_0", "rESR_RP"],
        full.df[full.df$drug_day=="RAPA_2", "rESR_RP"])




```

# Plot hsp70 microscopy timeries for fcz- and csp-treated cells ------------


## Import data

```{r}
datadir <- here("data", "microscopy", "csvs")
files <- list.files(datadir, full.names = TRUE, recursive = TRUE)


all.df.list <- lapply(files, function(x){
  df <- read.csv(x)
  df <- df[,-1]
  df[["channel"]] <- c("R", "G", "B")
  
  tb <- melt(df)

  filename <- unlist(lapply(strsplit(x, "/"), "[", 10))
  foldername <- unlist(lapply(strsplit(x, "/"), "[", 9))
  
  tb$day <- unlist(lapply(strsplit(foldername, "-"), "[", 3))
  tb$drug.series <- unlist(lapply(strsplit(foldername, "-"), "[", 2))
  image <- unlist(lapply(strsplit(filename, "-"), "[", 3))
  cell <- unlist(lapply(strsplit(as.character(tb$variable), "Area|Min|Max|Mean"), "[", 2))
  tb$image <- paste0(tb$day, tb$drug.series, image)
  tb$cell <- paste0(tb$image, cell, sep="_")
  tb$type <- "Area"
  tb$type[grep("Mean", tb$variable)] <- "Mean"
  tb$type[grep("Min", tb$variable)] <- "Min"
  tb$type[grep("Max", tb$variable)] <- "Max"
  
  tb$measure <- paste0(tb$type, tb$channel)
  return(tb)
})

all.df <- plyr::ldply(all.df.list)
```

```{r}
wide <- all.df %>%
  select(cell, drug.series, image, day, value, measure) %>%
  tidyr::pivot_wider(names_from = measure, values_from = value)
```


## Explore the two experimental batches
```{r}
ggplot(wide, aes(x=1:nrow(wide),y=MeanG, color = image)) +
  geom_point(size = 0.5) + 
  facet_grid(day~drug.series, scales = "free") +
  theme(legend.position = "none")

ggplot(wide, aes(x=1:nrow(wide),y=MeanR, color = image)) +
  geom_point() + 
  facet_grid(day~drug.series, scales = "free") +
  theme(legend.position = "none")
```
## Figure 3K

```{r}
make_bar <- function( N, intensity, bins=50, x_lab="mean intensity") {

  x <-  seq(0, max(N[[intensity]]), length.out=bins)
  N0 <- N %>% filter(day == "day0") 
  N2 <- N %>% filter(day == "day2") 
  N3 <- N %>% filter(day == "day3") 
  
  n0_log_idx <- c()
  for (i in 1:length(N0[[intensity]])) {
    n0_log_idx[i] <-min(which(N0[[intensity]][i] <= x))
  }
  split_n0 <- split(N0[[intensity]], n0_log_idx)
  count_n0 <- unlist(lapply(split_n0, FUN=function(x) return(length(x))))


  n2_log_idx <- c()
  for (i in 1:length(N2[[intensity]])) {
    n2_log_idx[i] <-min(which(N2[[intensity]][i] <= x))
  }
  split_n2 <- split(N2[[intensity]], n2_log_idx)
  count_n2 <- unlist(lapply(split_n2, FUN=function(x) return(length(x))))


  n3_log_idx <- c()
  for (i in 1:length(N3[[intensity]])) {
    n3_log_idx[i] <-min(which(N3[[intensity]][i] <= x))
  }
  split_n3 <- split(N3[[intensity]], n3_log_idx)
  count_n3 <- unlist(lapply(split_n3, FUN=function(x) return(length(x))))

  tmp <- tibble(day=c(rep(0,length(x)), rep(2,length(x)), rep(3,length(x))), binn=c(1:length(x), 1:length(x), 1:length(x)), count=0)
  
  for (i in 1:length(count_n0)) {
    idx <- which(tmp$day==0 & tmp$binn==names(count_n0)[i])
    tmp[idx, "count"] <- count_n0[i]/sum(count_n0)
  }

  for (i in 1:length(count_n2)) {
    idx <- which(tmp$day==2 & tmp$binn==names(count_n2)[i])
    tmp[idx, "count"] <- count_n2[i]/sum(count_n2)
  } 

  for (i in 1:length(count_n3)) {
    idx <- which(tmp$day==3 & tmp$binn==names(count_n3)[i])
    tmp[idx, "count"] <- count_n3[i]/sum(count_n3)
  } 
  tmp$day <- factor(tmp$day)
  g <- ggplot(tmp, aes(x=binn, y=count, fill=day)) +
    geom_bar(stat="identity", position=position_dodge()) +
    theme_bw() +
    ylim(c(0,0.08))+
    xlab("") +
    ylab("") +
    scale_fill_viridis(discrete = TRUE, direction = -1) +
    theme(panel.background = element_rect(fill = "white", colour = NA), 
        panel.border = element_blank(), 
        panel.grid = element_line(colour = "grey92"),
        panel.grid.minor = element_line(size = rel(0.5)), 
        strip.background = element_blank(),
        legend.key = element_rect(fill = "white", colour = NA), 
        legend.title = element_blank(), complete = TRUE)
  return(g)
  
}
```


```{r}
geldafcz <- make_bar(subset(wide, drug.series=="GeldaFCZ" & !day=="day0"), intensity="MeanG",  x_lab="Binned mean green intensity" )

fcz <- make_bar(subset(wide, drug.series=="FCZ" & !day=="day0"), intensity="MeanG",  x_lab="Binned mean green intensity" )

csp <- make_bar(subset(wide, drug.series=="CSP" & !day=="day0"), intensity="MeanG",  x_lab="Binned mean green intensity" )

library(ggpubr)
figure <- ggpubr::ggarrange(csp, fcz, geldafcz, nrow = 3, ncol = 1)

final.fig <- ggpubr::annotate_figure(figure,
                bottom = text_grob("Binned mean green intensity (n = 50 bins)"),
                left = text_grob("Frequency per bin", rot = 90)
                )
final.fig
ggsave("output/figure3k-hsp70_fczcspgelda_timeseries_microscopy.pdf", final.fig, height = 4, width = 6)
```


