---
title: "alevin preprocessing"
author: "Vanessa Dumeaux"
date: "04/15/2022"
output: html_document
---

# Load libraries and setup environment --------

```{r}
suppressPackageStartupMessages({
  library(R.utils)
  library(BiocParallel)
  library(here)
  library(dplyr)
})

ncores <- 24
register(MulticoreParam(workers=ncores))

source(here("src", "alevin_ca.R"))
source(here("src", "init.R"))


```

## paths
```{r}
genome_dir <-  here("data", "data01-preprocessing", "genome")
genome_path <-  here(genome_dir, "ca_genomic.fna")
gtf_path <-  here(genome_dir, "ca_original_genomic.gtf")
alevin_output <-  here("data", "data01-preprocessing", "sc-alevin", "alevin_output")
fastq_dir <- here("data", "data01-preprocessing", "fastq-input")

```

# Build genome index file ---------------
## build spliced transcriptome

We edited the make_splici_txome() function so it fits candida albicans genome annotations. The function can be found in `src/alevin_ca.R`

```{r}
read_length = 91
flank_trim_length = 5
output_dir = here(genome_dir, paste0("transcriptome_splici_fl", read_length - flank_trim_length))

make_splici_txome(gtf_path=gtf_path, 
                  genome_path=genome_path, 
                  read_length=read_length, 
                  flank_trim_length=flank_trim_length, 
                  output_dir=output_dir)
```
## Build Salmon/alevin index
```{r}
system(capture.output(
    cat(file.path(bin_folder,"salmon-1.6.0_linux_x86_64/bin/salmon"), "index -t",
        here(genome_dir, "transcriptome_splici_fl86", "transcriptome_splici_fl86.fa"),
        "-i", here(genome_dir,"transcriptome_splici_fl86", "ca_splici_idx"), "-p 16")))

```


# Align sc-reads and cell x gene quantification with Salmon/alevin-fry

You first need to download the raw fastq files from GEO. 

Then Salmon/alevin peforms selective alignment to generate a RAD file

```{r}

final <- readRDS(here("data", "final_sc_tibble_2.0.rds"))

target_names <- final$ids
target_names


alevin.oneliner <- function(target){
  system(capture.output(
    cat(file.path(bin_folder,"salmon-1.6.0_linux_x86_64/bin/salmon"), "alevin -i",
        here(genome_dir,"transcriptome_splici_fl86", "ca_splici_idx"), 
        "-p 16 -l IU --dropseq --sketch -1", file.path(fastq_dir, paste0(target, "_R1.fastq.gz")),
        "-2", file.path(fastq_dir, paste0(target, "_R2.fastq.gz")),
        "-o", file.path(alevin_output, paste0(target, "_map")), 
        "--tgMap", file.path(genome_dir, "transcriptome_splici_fl86", "transcriptome_splici_fl86_t2g.tsv"))
    ))
}

bplapply(target_names, alevin.oneliner) 
```

Alevin-fry uses that output to generate-permit list, collate the RAD file and perform quantification of the collated RAD file.

```{r}
alevin.fry <- function(target){
  system(capture.output(
    cat(file.path(anaconda_env_alevinfry, "alevin-fry"), "generate-permit-list -k -d fw -i", 
        file.path(alevin_output, paste0(target, "_map")),
        "-o", file.path(alevin_output, paste0(target, "_quant"))
    )
  ))
}

bplapply(target_names, alevin.fry) 

```
```{r}
alevin.fry.collate <- function(target){
  system(capture.output(
    cat(file.path(anaconda_env_alevinfry, "alevin-fry"), "collate -t 16 -i", 
        file.path(alevin_output, paste0(target, "_quant")), 
        "-r", file.path(alevin_output, paste0(target, "_map"))
    )
  ))
}
                  
bplapply(target_names, alevin.fry.collate) 
```

```{r}
alevin.fry.quant <- function(target){
  system(capture.output(
    cat(file.path(anaconda_env_alevinfry, "alevin-fry"), "quant -t 16 -i", 
        file.path(alevin_output, paste0(target, "_quant")), 
        "-o", file.path(alevin_output, paste0(target, "_quant_res")),
        "--tg-map", file.path(genome_dir, "transcriptome_splici_fl86", "transcriptome_splici_fl86_t2g_3col.tsv"),
        "--resolution cr-like --use-mtx")
  ))
}
                  
bplapply(target_names, alevin.fry.quant) 


```




















