---
title: "scRNA unfiltered pseudobulk"
author: "Vanessa Dumeaux"
date: "4/18/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# Load libraries and setup environment --------

```{r}
suppressPackageStartupMessages({
  library(R.utils)
  library(BiocParallel)
  library(here)
})

ncores <- 10
register(MulticoreParam(workers=ncores))

source(here("src", "init.R"))

final <- readRDS(here("data", "final_sc_tibble_2.0.rds"))

target_names <- final$ids
target_names

```

## paths
```{r}
genome_dir <-  here("data", "data01-preprocessing", "genome")
genome_path <-  here(genome_dir, "ca_genomic.fna")
gtf_path <-  here(genome_dir, "ca_original_genomic.gtf")
star_output <-  here("data", "data01-preprocessing", "pseudobulk-star", "star_output")
fastq_dir <- here("data", "data01-preprocessing", "fastq-input")

fastq_tagged <- here("data", "data01-preprocessing", "pseudobulk-star", "fastq_tagged")

```

You first need to download the raw fastq files from GEO

# Run droptag ----------

```{r}
if (!file.exists(fastq_tagged)){
  dir.create(fastq_tagged, showWarnings = FALSE) 
}

dropest.tag <- function(target){
  system(capture.output(
    cat(file.path(bin_folder, "local_dropest", "bin", "droptag"), "-c",
        here("data", "data01-preprocessing", "pseudobulk-star", "dropest_config", "drop_seq.xml"),
        "-S -s -p 8 -n", file.path(fastq_tagged, target), 
        "-l", file.path(fastq_tagged, target), 
        file.path(fastq_dir, paste0(target, "_R1.fastq.gz")),
        file.path(fastq_dir, paste0(target, "_R2.fastq.gz"))
        )
    ))
}


bplapply(target_names, dropest.tag)

```


# Build star index genome and align reads to genome -------

```{r}
system(capture.output(
    cat(file.path(bin_folder, "STAR-2.7.9a/bin/Linux_x86_64/STAR"), 
    "--runMode genomeGenerate --genomeDir", file.path(genome_dir, "star"),
    "--genomeFastaFiles", genome_path,
    "--sjdbGTFfile", gtf_path, "--sjdbOverhang 100")))

```

```{r}
if (!file.exists(star_output)){
  dir.create(star_output, showWarnings = FALSE) 
}

star.oneliner <- function(target){
  system(capture.output(
    cat("ls", file.path(fastq_tagged, paste0(target, ".fastq.gz")),"| paste -sd \",\" - | gawk \'{ print", paste0("\"", file.path(bin_folder, "STAR-2.7.9a/bin/Linux_x86_64/STAR")), "--genomeDir", file.path(genome_dir, "star"), "--runThreadN 8 --readFilesIn \" $0 \" --outFileNamePrefix",
        file.path(star_output, paste0(target,".out")), "--readFilesCommand zcat --outSAMmultNmax 1 --outSAMunmapped Within --outSAMtype BAM SortedByCoordinate\"}\' | bash")
))
}

bplapply(target_names, star.oneliner) 

```

# Build pseudobulk gene x sample count matrix

```{r}
library(Rsubread)

files <- list.files(star_output, pattern = "Aligned.sortedByCoord.out.bam$", full.names = TRUE)

f1 <- featureCounts(files=files,
                    annot.ext = here(genome_dir, "ca_original_genomic.gtf"),
                    isGTFAnnotationFile = TRUE, 
                    strandSpecific = 1, #stranded
                    GTF.attrType = "gene_id",
                    # fractional counting of multimapping reads
                    countMultiMappingReads = TRUE,
                    fraction = TRUE,
                    nthreads = 30)


saveRDS(f1, here("data", "data01-preprocessing", "pseudobulk-star", "pseudobulk_counts_geneid.rds"))

```

