---
title: "bulk RNAseq preprocessing using STAR"
author: "Vanessa Dumeaux"
date: "04/18/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Load libraries and setup environment --------

```{r}
suppressPackageStartupMessages({
  library(R.utils)
  library(BiocParallel)
  library(here)
})

ncores <- 24
register(MulticoreParam(workers=ncores))

source(here("src", "init.R"))

fastq_dir <- here("data", "data01-preprocessing", "fastq_input")


genome_dir <-  here("data", "data01-preprocessing", "genome")

```

You first need to download the raw fastq files `FCZ_2_bulk_R1.fastq.gz` and `FCZ_2_bulk_R2.fastq.gz` from GEO


# Trim reads -----------

```{r, trimmed filenames, message=FALSE, warning=FALSE}
trimmed <- here("data", "data01-preprocessing", "bulk-star", "trimgalore_output")
if (!file.exists(trimmed)){
  dir.create(trimmed, showWarnings = FALSE) 
}


system(capture.output(
  cat(file.path(bin_folder, "TrimGalore-0.6.6/trim_galore"),
      "--path_to_cutadapt",  file.path(anaconda_env_cutadapt, "cutadapt"),
      "-j 4 --paired --retain_unpaired --phred33 --dont_gzip -o", trimmed,
      "--length 14 --stringency 1 -e 0.1 --2colour 20",
      file.path(fastq_dir, "FCZ_2_bulk_R1.fastq.gz"), 
      file.path(fastq_dir, "FCZ_2_bulk_R2.fastq.gz"))
    ))


```

## Align using STAR

The genome index for star has already been created by code in `02-scrna-unfiltered-pseudobulk-star.Rmd`

```{r, aligning filenames, message=FALSE, warning=FALSE}

aligned <- here("data", "data01-preprocessing", "bulk-star", "star_output")
if (!file.exists(aligned)){
  dir.create(aligned, showWarnings = FALSE) 
}


system(capture.output(
    cat(file.path(bin_folder, "STAR-2.7.9a/bin/Linux_x86_64/STAR"),
    "--genomeDir", file.path(genome_dir, "star"), 
    "--runThreadN 20 --readFilesIn", file.path(trimmed, "FCZ_2_bulk_R1_val_1.fq"), file.path(trimmed, "FCZ_2_bulk_R2_val_2.fq"),
    "--outSAMmultNmax 1 --outSAMunmapped Within --outSAMtype BAM SortedByCoordinate --outSAMattributes All --outSAMstrandField intronMotif --quantMode GeneCounts --outFileNamePrefix", file.path(aligned, "FCZ_2_bulk"))
  ))


```


```{r}
library(Rsubread)

file <- file.path(aligned, paste0("FCZ_2_bulk", "Aligned.sortedByCoord.out.bam"))

f1 <- featureCounts(files=file,
                    annot.ext = file.path(genome_dir, "ca_original_genomic.gtf"),
                    isGTFAnnotationFile = TRUE, 
                    strandSpecific = 1, #stranded
                    GTF.attrType = "gene_id",
                    # parameters specific to paired end reads
                    isPairedEnd = TRUE,
                    countReadPairs = TRUE,
                    requireBothEndsMapped = TRUE,
                    checkFragLength = FALSE,
                    countChimericFragments = FALSE,    
                    autosort = TRUE,
                    # fractional counting of multimapping reads
                    countMultiMappingReads = TRUE,
                    fraction = TRUE,
                    nthreads = 30)


saveRDS(f1, here("data", "data01-preprocessing", "bulk-star", "bulk_counts_geneid.rds"))

```

